Fix NullReferenceException in FileSystemWatcher on Windows during dispose

martincostello opened the pull request:
Fix NullReferenceException in FileSystemWatcher on Windows if two threads try and dispose the instance at the same time.

Found in a failing test while running the CI in a PR against the OpenTelemetry libraries for net8.0 on Windows: [workflow logs](https://github.com/open-telemetry/opentelemetry-dotnet/actions/runs/8600318104/job/23565066419?pr=5520#step:6:3896)

```console
Failed OpenTelemetry.Instrumentation.AspNetCore.Tests.BasicTests.AddAspNetCoreInstrumentation_BadArgs [1 ms]
  Error Message:
   [Test Class Cleanup Failure (OpenTelemetry.Instrumentation.AspNetCore.Tests.BasicTests)]: System.NullReferenceException : Object reference not set to an instance of an object.
  Stack Trace:
     at System.IO.FileSystemWatcher.StopRaisingEvents()
   at System.IO.FileSystemWatcher.Dispose(Boolean disposing)
   at System.ComponentModel.Component.Dispose()
   at Microsoft.Extensions.FileProviders.Physical.PhysicalFilesWatcher.Dispose(Boolean disposing)
   at Microsoft.Extensions.FileProviders.PhysicalFileProvider.Dispose(Boolean disposing)
   at Microsoft.Extensions.FileProviders.PhysicalFileProvider.Dispose()
   at Microsoft.Extensions.Hosting.Internal.Host.<DisposeAsync>g__DisposeAsync|21_0(Object o)
   at Microsoft.Extensions.Hosting.Internal.Host.DisposeAsync()
   at Microsoft.Extensions.Hosting.Internal.Host.Dispose()
   at Microsoft.AspNetCore.Mvc.Testing.DeferredHostBuilder.DeferredHost.Dispose()
   at Microsoft.AspNetCore.Mvc.Testing.WebApplicationFactory`1.DisposeAsync()
   at Microsoft.AspNetCore.Mvc.Testing.WebApplicationFactory`1.DisposeAsync()
   at Microsoft.AspNetCore.Mvc.Testing.WebApplicationFactory`1.Dispose(Boolean disposing)
   at Microsoft.AspNetCore.Mvc.Testing.WebApplicationFactory`1.Dispose()
```

stephentoub commented:
> if two threads try and dispose the instance at the same time

It's not thread-safe. If two threads are calling Dispose concurrently, that's a bug in the consumer.

martincostello commented:
Sure, but trying to find where the exception was coming from it looks like both the Linux and macOS implementations of `StopRaisingEvents()` guard against the state changing to a degree:

https://github.com/dotnet/runtime/blob/70e3218806e88465043a25d50bf7c2305aa2177d/src/libraries/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.OSX.cs#L69-L76

https://github.com/dotnet/runtime/blob/70e3218806e88465043a25d50bf7c2305aa2177d/src/libraries/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.Linux.cs#L98-L103

Windows seems to be the odd-one-out here, so seemed reasonable to me to add the null guard.

I haven't looked into OTel's tests as to why its happening - it could be an issue with their tests, it could be something in xunit as it appears to happen during test class cleanup.

martincostello commented:
Looks like [a standard xunit `IClassFixture<T>` usage](https://github.com/open-telemetry/opentelemetry-dotnet/blob/e46a446c4a8a850b024645f697fbb82db602358a/test/OpenTelemetry.Instrumentation.AspNetCore.Tests/BasicTests.cs#L24-L34), which itself is copied from [some old samples in aspnetcore](https://github.com/dotnet/AspNetCore.Docs/blob/c429badd4799dfbbddb1c67cdc5e85cd704d13a7/aspnetcore/test/integration-tests/samples/3.x/IntegrationTestsSample/tests/RazorPagesProject.Tests/IntegrationTests/BasicTests.cs#L8-L16).

stephentoub commented:
> both the Linux and macOS implementations of StopRaisingEvents() guard against the state changing to a degree:

If the goal is consistency, then it'd be better to do:
```C#
SafeFileHandle? handle = _directoryHandle;
if (handle is not null)
{
    _directoryHandle = null;
    handle.Dispose();
}
```
That won't address other problems with this instance being erroneously used concurrently, though.

martincostello commented:
Sure, I'll refactor that then. I went with `?` as it was the smallest change.

martincostello committed:
Fix NullReferenceException

Fix `NullReferenceException` in `FileSystemWatcher` on Windows if two threads try and dispose the instance at the same time.
